### DISCUSSION TITLE
Auto import not working for symbols in`__init__.py`

### DISCUSSION THREAD

**Discussion by daya0576** [microsoft/pyright#4018] [Q&A] [answered]
I understand that auto import only works for symbols that have been imported. 

But for the symbol `foo` already imported in file `a.py`, still can not be found in `b.py`:

```
├── a.py
├── b.py
└── module
    └── __init__.py

==> __init__.py <==
class Foo:
    ...
foo = Foo()
__all__ = ["foo"]
```

```python
==> a.py <==
# import symbol foo manually
from module import foo
print(foo)

==> b.py <==
# auto import still not working
foo

```

**Comment by erictraut** [comment, marked as answer]
Yes, this is the intended behavior in pyright. Only symbols that have already been seen by the type checker and are cached in memory are offered as auto-import suggestions.

Pylance, Microsoft's full-featured language server for Python, offers full indexing capabilities for symbols in your project, so it's able to offer auto-import suggestions even when they are not yet used.

**Reply by daya0576** [reply]
> Only symbols that have already been seen by the type checker and are cached in memory are offered as auto-import suggestions.

But in the file `a.py`, symbol `foo` should be imported and cached, but it still can not be seen in file `b.py`.

p.s the `diagnosticMode` option has been set to workspace.

**Reply by erictraut** [reply]
You're correct. I would expect this to work. 

This is due to [this code](https://github.com/microsoft/pyright/blob/f21d04f204d1cccdf5adc11d740c34e31b7a6bdb/packages/pyright-internal/src/languageService/autoImporter.ts#L470) in the auto importer logic. The code is including only constants or type aliases, not variables. A few lines earlier, the code allows variables that are included in a `__all__`, as it is in your example. However, that code is conditionalized based on an options field `allowVariableInAll` which is disabled by default.

This code is part of the pyright code base, but it's maintained by the pylance team, since it's a language server feature and not part of the core type checker. I'm going to convert this thread to a bug report and then transfer it to the pylance-release project so it gets the appropriate attention.

**Reply by daya0576** [reply]
Got it, thx for your detailed explanation.

Now is it possible to enable `allowVariableInAll` locally? It seems unlikely according to [this code](https://github.com/microsoft/pyright/blob/f21d04f204d1cccdf5adc11d740c34e31b7a6bdb/packages/pyright-internal/src/languageService/completionProvider.ts#L2219).

**Reply by erictraut** [reply]
No, it's not possible to enable this flag externally to the program. It's an internal implementation detail.
